### PR TITLE
Device association

### PR DESCRIPTION
**Pull request by pierredelisle**

### CODE REVIEW

File: doc/DeviceAssociationInLocalNetwork.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
+# Device Association in Local Network
```
Comment (CodeChronos928) [inline doc/DeviceAssociationInLocalNetwork.md:1]: General notes for the docs I fixed in my previous commits:

- Titles / Headings should be in sentence case, according to the TW authorities.  I was very resistant to this for a long time, but now just do it automatically.
- Please wrap the paragraphs at 65 characters.  It doesn't affect how they render, but makes them more readable / editable in a text editor.

File: doc/DeviceAssociationInLocalNetwork.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+NOTE: Support for Device Association between the lock and ocsensor example applications
+is not yet complete.
+
+Note that the multicast message will not work because of Issue
+[Multicast ipv6 addresses for sleepy end devices not properly supported #598](https://github.com/openweave/openweave-core/issues/598).
+
+### Weave artifacts supporting device association
+
+##### DeviceDescription
```
Comment (CodeChronos928) [inline doc/DeviceAssociationInLocalNetwork.md:26]: If this is a variable, it should be in back-ticks.  E.g:   `DeviceDescription`

File: doc/DeviceAssociationInLocalNetwork.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+1. lock2 triggers an association request
+1. sensor1 is still in ”association-ready” mode and replies to lock2’s association request
+1. lock2 and sensor1 are associated. Wrong!
+
+To help mitigate the issue of sensor1 being associated with two locks (as is the case with 
+steps 6 and 7), the following is put in place:
+- When sensor1 is placed in “association-ready” mode (step 2), there is a timeout that will cancel that mode when it goes off (e.g. timeout value of 1 minute).
+- When sensor1 replies to an association request (step 3), the timeout value is immediately reduced to the lowest  of X seconds (e.g. 5 seconds) and the timer’s residual value to help reduce the likelihood that step 6 will occur. This way, there is still a short period of a few seconds that allows to account for communication failures and retries, but this is a shorter period than what might have been left on the residual timeout period, lowering the likelihood that sensor1 is still in “association-ready” mode when step 5 occurs.
+
+FIXME: To be implemented.
```
Comment (CodeChronos928) [inline doc/DeviceAssociationInLocalNetwork.md:105]: Nit:  I might call-out FIXMEs with Markdown "note" syntax.  E.g.

> FIXME: Fix this

File: src/examples/lock/DeviceController.cpp [hunk 4]
```diff
@@ -433,37 +436,56 @@ void DeviceController::FactoryResetButtonHandler()
...
+    char addrStr[50];
+    allThreadNodesAddr.ToString(addrStr, sizeof(addrStr));
+    WeaveLogDetail(Support, "All thread nodes multicast address: [%s]", addrStr);
+
+    IdentifyRequestMessage identifyReqMsg;
     identifyReqMsg.TargetFabricId  = ::nl::Weave::DeviceLayer::FabricState.FabricId;
     identifyReqMsg.TargetModes     = kTargetDeviceMode_UserSelectedMode;
-    identifyReqMsg.TargetVendorId  = vendorId;
-    identifyReqMsg.TargetProductId = productId;
+    identifyReqMsg.TargetVendorId  = 0xFFFF; // Any vendor
```
Comment (CodeChronos928) [inline src/examples/lock/DeviceController.cpp:478]: Is using 0xFFFF for VID/PID an accepted practice, and documented in an RFC somewhere?  If so, please call it out.

File: src/examples/lock/DeviceController.cpp [hunk 5]
```diff
@@ -433,37 +436,56 @@ void DeviceController::FactoryResetButtonHandler()
...
+    identifyReqMsg.TargetProductId = 0xFFFF; // Any product
     identifyReqMsg.TargetDeviceId  = nl::Weave::kAnyNodeId;
 
     WeaveLogProgress(Support, "Sending the Identify request");
-    err = _this.mDeviceDescriptionClient.SendIdentifyRequest(ip_addr, identifyReqMsg);
+    WeaveLogProgress(Support,
+            "fabric [0x%016" PRIx64 "] modes [0x%08" PRIx32 "] vendor [0x%04" PRIx16 "] product [0x%04" PRIx16 "] device [0x%016" PRIx64 "]",
```
Comment (CodeChronos928) [inline src/examples/lock/DeviceController.cpp:484]: Thank you for not making me tell you to use PRIx64, or explain how to.  :)

File: src/examples/lock/DeviceController.cpp [hunk 6]
```diff
@@ -35,6 +36,8 @@
 #include "WDMFeature.h"
 #include "AppTask.h"
 
+#include <inttypes.h>
```
Comment (CodeChronos928) [inline src/examples/lock/DeviceController.cpp:39]: This looks like the correct header ordering, but just want to confirm we're running these through the Google linter to enforce stuff like that.

**Review by CodeChronos928 — COMMENTED** [review]
Comments added.

File: src/examples/lock/DeviceController.cpp (hunk 6 above)
Comment (pierredelisle) [reply src/examples/lock/DeviceController.cpp:39]: Did not pay much attention about the ordering. Feel free to file an issue so we address it in the future.

File: src/examples/lock/DeviceController.cpp (hunk 4 above)
Comment (pierredelisle) [reply src/examples/lock/DeviceController.cpp:478]: It comes from DeviceDescription.cpp. We could file a bug for openweave to define a constant for it.

**Pull request merged**